### PR TITLE
Service now exporter

### LINKED ISSUE #126: Extend failure exporter to get metrics from ServiceNow
Enhance failure exporter to gather metrics from ServiceNow.

### PR DESCRIPTION
**Pull request by malacourse**
## Describe the behavior changes introduced in this PR
This PR adds support for exporting ServiceNow incidents or problems using the ServcieNow API.  This will create data for failure time resolution.

## Linked Issues?

resolves #126 

## Testing Instructions
With a developer account you can create a ServiceNow instance for testing: https://developer.servicenow.com/dev.do

1) Start the service now instance
2) Create a field to be used as the Application Label to match labels in the OCP deployments.
  a) When viewing an incident use the upper left hamburg menu and select "Configure -> Form Layout"
  b) Create a custom field of type string or list to be used for the application field
  c) Add the field to the form layout
2) Create some resolved incidents and populate the App Label field with values
3) set your environment variables similar to the following:
      SERVER='https://dev59871.service-now.com'
      TRACKER_QUERY='/api/now/table/incident?sysparm_fields=closed_at%2Copened_at%2Cstate%2Cnumber%2Cu_applabel&sysparm_display_value=true&state=6'
      USER=admin
      TOKEN=<password>
      TRACKER_PROVIDER=servicenow
      APP_LABEL_FIELD=u_applabel

Please include any additional commands or pointers in addition to our [standard PR testing process](/docs/Development.md#testing-pull-requests).

@redhat-cop/mdt

### CODE REVIEW

File: docs/Configuration.md [hunk 1]
```diff
@@ -156,8 +156,10 @@ This exporter provides several configuration options, passed via environment var
 
 | Variable | Required | Explanation | Default Value |
 |---|---|---|---|
+| `TACKER_PROVIDER` | no | Set the type of failure provider. One of `jira`, `servicenow` | `jira` |
```
Comment (etsauer) [inline docs/Configuration.md:159]: Let's just call this `PROVIDER` to be consistent with other exporters

File: docs/Configuration.md [hunk 2]
```diff
@@ -156,8 +156,10 @@ This exporter provides several configuration options, passed via environment var
...
 | `LOG_LEVEL` | no | Set the log level. One of `DEBUG`, `INFO`, `WARNING`, `ERROR` | `INFO` |
-| `SERVER` | yes | URL to the Jira Server  | unset  |
-| `PROJECT` | yes | Jira project to scan | unset |
-| `USER` | yes | Jira Username | unset |
+| `SERVER` | yes | URL to the Jira or ServiceNowServer  | unset  |
+| `PROJECT` | yes | Jira project or ServiceNow category to scan | unset |
+| `USER` | yes | Tracker Username | unset |
 | `TOKEN` | yes | User's API Token | unset |
+| `TRACKER_QUERY` | no | Used for ServiceNow integration, query used to gather incidents or problems. ex: "/api/now/table/incident?sysparm_fields=closed_at%2Copened_at%2Cstate%2Cnumber&state=6&category=" | unset |
```
Comment (etsauer) [inline docs/Configuration.md:165]: let's call this `INCIDENT_QUERY`

**Review by etsauer — CHANGES_REQUESTED** [review]
@malacourse made a couple comments on the variable names. In addition two major items that need addressed>

1. When I run this I get the following error:

```
Exception has occurred: TypeError
__init__() takes 4 positional arguments but 5 were given
  File "/home/esauer/src/pelorus/exporters/failure/collector_servicenow.py", line 22, in __init__
    super().__init__(server, user, apikey, project)
  File "/home/esauer/src/pelorus/exporters/failure/app.py", line 21, in getCollector
    return ServiceNowFailureCollector(username, token, tracker_api, project)
  File "/home/esauer/src/pelorus/exporters/failure/app.py", line 38, in <module>
    collector = TrackerFactory.getCollector(username, token, tracker_api, project, tracker_provider)
```

2. We need a method of figuring out the app name related to a ticket. See the updates made to the Jira collector for a reference.

File: exporters/failure/collector_servicenow.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            logging.info('Found issue opened: %s, %s: %s', issue.get('number'),
+                         issue.get('opened_at'), issue.get('closed_at'))
+            # Create the JiraFailureMetric
+            created_ts = self.convert_timestamp(issue['opened_at'])
+            resolution_ts = None
+            if issue['closed_at']:
+                logging.info('Found issue close: %s, %s: %s', issue.get('closed_at'),
+                             issue.get('number'), issue.get('opened_at'))
+                resolution_ts = self.convert_timestamp(issue.get('closed_at'))
+            tracker_issue = TrackerIssue(issue.get('number'), created_ts, resolution_ts)
```
Comment (etsauer) [inline exporters/failure/collector_servicenow.py:52]: This line needs to include an app name, like so:

```
tracker_issue = TrackerIssue(issue.key, created_ts, resolution_ts, self.get_app_name(issue))
```

**Review by etsauer — CHANGES_REQUESTED** [review]
ok, so a couple of things here.... it looks like there are several different sys_params for "opening/creating" and "resolving/closing" an incident. Are we using the right ones?

second, I think we need to get rid of `TRACKER_QUERY` and instead build the query ourselves using the APP_LABEL parameter. This is both because then it works more the way the Jira provider does, AND because it makes me a little worried about security holes we might be opening by having the user define a custom query.

also, in the current verison, you need to pass the value of `APP_LABEL` twice, once in the variable and once in the query, which I think will be confusing

File: docs/Configuration.md [hunk 4]
```diff
@@ -156,8 +156,11 @@ This exporter provides several configuration options, passed via environment var
...
+| `APP_LABEL_FIELD` | yes | ServiceNow field used for the Application label. ex: "u_applabel" | unset |
```
Comment (etsauer) [inline docs/Configuration.md:166]: This should just be `APP_LABEL`

**Comment by malacourse** [comment]
Changes have been pushed per review

**Comment by malacourse** [comment]
I realized I pushed to wrong branch.  Updates are now in.

**Comment by etsauer** [comment]
@malacourse so it looks like two things are happening now when I call the exporter.

First it appears to be returning the same results multiple times per call:

```
$ curl localhost:8080 | grep 'INC0010010'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_resolution_timestamp{app="unknown",issue_number="INC0010010"} 1.603796262e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
100  2failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
51k  10failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
0  251failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
k    0     0   118kfailure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
      0 failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
 0:00:02  failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
0:00:02 failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
--:--:-failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
-  118k
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
failure_creation_timestamp{app="unknown",issue_number="INC0010010"} 1.603796236e+09
```

Second, the incident that i searched for above should have been labelled `app=basic-spring-boot` per the screenshot below, and it wasn't.

inc-spring-boot

**Review by etsauer — APPROVED** [review]
lgtm

**Pull request merged**